server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/DefaultMetaRepository.java

```diff
@@ -68,307 +55,147 @@ public class DefaultMetaRepository extends RepositoryWrapper implements MetaRepo
-        final Map<String, Entry<?>> entries =
-                find(new Revision(rev), PATH_CREDENTIALS_AND_MIRRORS, Collections.emptyMap()).join();
+    private CompletableFuture<Map.Entry<List<Mirror>, List<MirrorCredential>>> mirrorsAndCredentials(int rev) {
+        return find(new Revision(rev), PATH_CREDENTIALS_AND_MIRRORS, Collections.emptyMap()).thenApply(
+                entries -> {
+                    if (entries.isEmpty()) {
+                        return EMPTY_MIRRORS_AND_CREDENTIALS;
+                    }
+
+                    try {
+                        final List<MirrorCredential> credentials = parseCredentials(entries);
+                        final List<Mirror> mirrors = parseMirrors(entries, credentials);
+                        return Maps.immutableEntry(mirrors, credentials);
+                    } catch (RepositoryMetadataException e) {
+                        throw e;
+                    } catch (Exception e) {
+                        throw new RepositoryMetadataException("failed to load the mirror configuration", e);
+                    }
+                });
+    }
 
-        if (!entries.containsKey(PATH_MIRRORS)) {
-            return Collections.emptySet();
+    private List<Mirror> parseMirrors(Map<String, Entry<?>> entries, List<MirrorCredential> credentials)
+            throws JsonProcessingException {
+        final Entry<?> e = entries.get(PATH_MIRRORS);
+        if (e == null) {
+            return ImmutableList.of();
         }
 
-        final JsonNode mirrorsJson = (JsonNode) entries.get(PATH_MIRRORS).content();
+        final JsonNode mirrorsJson = (JsonNode) e.content();
         if (!mirrorsJson.isArray()) {
             throw new RepositoryMetadataException(
                     PATH_MIRRORS + " must be an array: " + mirrorsJson.getNodeType());
         }
-
-        if (mirrorsJson.size() == 0) {
-            return Collections.emptySet();
+        if (mirrorsJson.isEmpty()) {
+            return ImmutableList.of();
         }
 
-        try {
-            final List<MirrorCredential> credentials = loadCredentials(entries);
-            final ImmutableSet.Builder<Mirror> mirrors = ImmutableSet.builder();
+        final ImmutableList.Builder<Mirror> mirrors = ImmutableList.builder();
 
-            for (JsonNode m : mirrorsJson) {
-                final MirrorConfig c = Jackson.treeToValue(m, MirrorConfig.class);
-                if (c == null) {
-                    throw new RepositoryMetadataException(PATH_MIRRORS + " contains null.");
-                }
-                mirrors.addAll(c.toMirrors(parent(), credentials));
+        int index = 0;
+        for (JsonNode m : mirrorsJson) {
+            final MirrorConfig c = Jackson.treeToValue(m, MirrorConfig.class);
+            if (c == null) {
+                throw new RepositoryMetadataException(PATH_MIRRORS + " contains null.");
+            }
+            final Mirror mirror = c.toMirror(parent(), credentials, index++);
```

Comment:
Question) What do you think of just persistently assigning an id? I understand that previously we didn't have mirrors, but I think it's possible to migrate to an object based schema instead of a mirror based schema and enforce ids and create an arbitrary id if missing.

With the current setup, I fear that users might make a mistake if multiple users modify mirror for the same project. i.e. Two users may try to modify a mirror configuration concurrently, and the index mapping might change.